File: app/src/main/java/org/carstenf/wordfinder/WordFinderSettingsFragment.java
package org.carstenf.wordfinder;

import android.os.Bundle;
import android.text.InputFilter;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceFragmentCompat;

import com.takisoft.preferencex.EditTextPreference;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class WordFinderSettingsFragment extends PreferenceFragmentCompat {
	@Override
	public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
		setPreferencesFromResource(R.xml.preferences, rootKey);
		EditTextPreference timeEdit = findPreference("countdown_time_pref");

		if(timeEdit!=null)
			timeEdit.setOnBindEditTextListener( (@NonNull EditText editText) -> {
					List<InputFilter> filters = new ArrayList<>(Arrays.asList(editText.getFilters()));
					filters.add((CharSequence source, int start, int end, Spanned dest, int dstart,
								 int dend) -> {
							SpannableStringBuilder builder = new SpannableStringBuilder(dest);
							builder.replace(dstart, dend, source, start, end);
							if( builder.toString().matches("[0-9]+(:(|[0-5]|[0-5][0-9]))?")) //     :?( [ ?[0-5]?[0-9]?"))
								return null;
							return "";
					});
					editText.setFilters(filters.toArray(new InputFilter[0]));
			});
	}
}

File: app/src/main/kotlin/org/carstenf/wordfinder/WordFinderSettingsFragment.kt
/*
 * Copyright Carsten Friedrich (Carsten.Friedrich@gmail.com)
 *
 * License: GNU GENERAL PUBLIC LICENSE 3.0 (https://www.gnu.org/copyleft/gpl.html)
 *
 */
package org.carstenf.wordfinder

import android.os.Bundle
import android.text.InputFilter
import android.text.SpannableStringBuilder
import androidx.preference.PreferenceFragmentCompat
import com.takisoft.preferencex.EditTextPreference

class WordFinderSettingsFragment : PreferenceFragmentCompat() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)

        val timePref: EditTextPreference? = findPreference("countdown_time_pref") // NON-NLS
        if(timePref!=null) {
            timePref.setSummary(
                String.format(
                    getString(R.string.PrefTimeLimitSummary),
                    timePref.text
                )
            )

            timePref.setOnPreferenceChangeListener{_, newValue ->
                timePref.setSummary(
                    String.format(
                        getString(R.string.PrefTimeLimitSummary),
                        newValue
                    )
                )
                true
            }

            timePref.setOnBindEditTextListener { editText ->
                val filters = editText.filters.toMutableList()
                filters.add(InputFilter { source, start, end, dest, dstart, dend ->
                    val builder = SpannableStringBuilder(dest)
                    builder.replace(dstart, dend, source, start, end)
                    if (builder.toString().matches(Regex("[0-9]+(:([0-5]|[0-5][0-9])?)?"))) {
                        null
                    } else {
                        ""
                    }
                })
                editText.filters = filters.toTypedArray()
            }
        }
    }
}
